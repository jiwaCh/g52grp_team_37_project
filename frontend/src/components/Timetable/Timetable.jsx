import React from 'react';
import proptypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { Thunks as TimetableActions } from '../../store/timetable';
import { List } from '@material-ui/core';
import TimetableItem from '../TimetableItem/TimetableItem';
import useStyles from './Timetable.styles';

/**
 * Timetable component
 *
 * @param {{id: string}} props props.id is the bus stop id
 * @returns Timetable Component
 */
const Timetable = ({ id }) => {
  const { timetable, loading, busStopId, retrievedAt } = useSelector(state => state.timetable);
  const dispatch = useDispatch();

  const expiration = new Date(Date.now() - 60 * 1000);

  if (!loading && (busStopId !== id || retrievedAt < expiration)) {
    dispatch(TimetableActions.fetchTimetable(id));
  }

  const classes = useStyles();

  return (
    <>
      <List className={classes.root}>
        {loading && 'Loading...'}
        {!loading &&
          timetable.map((entry, i) => (
            <TimetableItem
              key={entry.busId}
              busLine={entry.busLine}
              destination={entry.destination}
              time={entry.arrivalTime}
              color={entry.color}
              alternate={i % 2}
            />
          ))}
      </List>
      <div className={classes.fixedBottom} />
    </>
  );
};

Timetable.propTypes = {
  id: proptypes.string.isRequired,
};

export default Timetable;
