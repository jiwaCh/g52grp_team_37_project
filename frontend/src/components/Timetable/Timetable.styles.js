import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  root: {
    bottom: 0,
    overflow: 'scroll',
  },
}));

export default useStyles;
