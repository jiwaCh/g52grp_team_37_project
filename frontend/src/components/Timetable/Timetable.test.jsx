import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Timetable from './Timetable';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

const mockStore = configureMockStore([thunk]);

it('renders correctly without state', () => {
  const store = mockStore({
    timetable: {
      loading: false,
      timetable: [],
      busStopId: '',
      retrievedAt: undefined,
      error: undefined,
    },
  });

  const component = shallow(
    <Provider store={store}>
      <Timetable id="some id" />
    </Provider>,
  );

  expect(toJson(component)).toMatchSnapshot();
});

it('renders correctly when loading', () => {
  const store = mockStore({
    timetable: {
      loading: true,
      timetable: [],
      busStopId: '',
      retrievedAt: undefined,
      error: undefined,
    },
  });

  const component = shallow(
    <Provider store={store}>
      <Timetable id="some id" />
    </Provider>,
  );

  expect(toJson(component)).toMatchSnapshot();
});

it('renders correctly when loaded', () => {
  const store = mockStore({
    timetable: {
      loading: false,
      timetable: [
        { busId: '1234', busLine: '31', arrivalTime: '21:30', destination: 'Victoria Centre', color: '#DA487E' },
        { busId: '1235', busLine: '903', arrivalTime: '21:45', destination: 'University Park', color: '#18194f' },
        { busId: '1236', busLine: '34', arrivalTime: '22:00', destination: 'City Loop', color: '#E37222' },
      ],
      busStopId: '',
      retrievedAt: undefined,
      error: undefined,
    },
  });

  const component = shallow(
    <Provider store={store}>
      <Timetable id="some id" />
    </Provider>,
  );

  expect(toJson(component)).toMatchSnapshot();
});
