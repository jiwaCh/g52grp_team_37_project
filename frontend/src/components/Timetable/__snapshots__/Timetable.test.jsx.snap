// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when loaded 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Timetable
    id="some id"
  />
</ContextProvider>
`;

exports[`renders correctly when loading 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Timetable
    id="some id"
  />
</ContextProvider>
`;

exports[`renders correctly without state 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Timetable
    id="some id"
  />
</ContextProvider>
`;
